server/verification-request/verification-request.service.ts

```diff
@@ -557,4 +889,472 @@ export class VerificationRequestService {
+        this.logger.error(`Error in state ${state} for VR ${vrId}: ${error}`);
+
+        await this.VerificationRequestModel.findByIdAndUpdate(vrId, {
+            $push: {
+                stateErrors: {
+                    state,
+                    error,
+                    timestamp: new Date(),
+                },
+            },
+        });
+    }
+
+    /**
+     * Track state transition
+     */
+    private async trackStateTransition(
+        vr: VerificationRequestDocument,
+        from: string,
+        to: string,
+        duration: number
+    ) {
+        await this.VerificationRequestModel.findByIdAndUpdate(vr._id, {
+            $push: {
+                stateTransitions: {
+                    from,
+                    to,
+                    timestamp: new Date(),
+                    duration,
+                },
+            },
+        });
+
+        this.logger.log(
+            `State transition: ${from} -> ${to} in ${duration}ms for VR ${vr.id}`
+        );
+    }
+
+    /**
+     * Update progress tracking
+     */
+    private async updateProgress(vr: VerificationRequestDocument) {
+        const statesExecuted = vr.statesExecuted || [];
+        const totalStates = EXPECTED_STATES.length;
+        const completed = statesExecuted.length;
+        const percentage = (completed / totalStates) * 100;
+
+        const avgDuration = this.calculateAverageDuration(
+            vr.stateTransitions || []
+        );
+        const remainingStates = totalStates - completed;
+        const estimatedCompletion =
+            avgDuration > 0
+                ? new Date(Date.now() + avgDuration * remainingStates)
+                : undefined;
+
+        await this.VerificationRequestModel.findByIdAndUpdate(vr._id, {
+            progress: {
+                current:
+                    statesExecuted[statesExecuted.length - 1] || "starting",
+                completed,
+                total: totalStates,
+                percentage,
+                estimatedCompletion,
+            },
+        });
+    }
+
+    /**
+     * Calculate average duration from transitions
+     */
+    private calculateAverageDuration(transitions: any[]): number {
+        if (!transitions || transitions.length === 0) return 0;
+
+        const total = transitions.reduce(
+            (sum, t) => sum + (t.duration || 0),
+            0
+        );
+        return total / transitions.length;
+    }
+
+    /**
+     * Revalidate and run missing states with parallel execution
+     * IMPORTANT: Topics and Impact Area can only run AFTER identifiedData is populated
+     *
+     * This function checks pending tasks to prevent duplicate AI task creation.
+     * When identifiedData completes, it triggers topics and impactArea in parallel.
+     * It avoids falling back to sequential execution if these tasks are already pending.
+     */
+    private async revalidateAndRunMissingStatesWithParallel(
```

Comment:
question: The state machine guard conditions assume sequential execution, but can this method triggers multiple states in parallel ?

Reply:
If the validation works, yes